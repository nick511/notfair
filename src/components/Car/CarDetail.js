import React, { Component } from 'react'
import PropTypes from 'prop-types'
import cn from 'classnames'

import styles from './CarDetail.module.scss'

export class CarDetail extends Component {
  _favoriteCar = () => {
    const { favoriteCarAction, car } = this.props
    favoriteCarAction(car.id, !car.isFavorite)
  }

  render() {
    const { car } = this.props
    const payment = car.product_financials[0]

    return (
      <div className={styles.component}>
        {/* Favorite icon */}
        <div
          className={cn(styles.favorite, 'favorite', {
            favorited: car.isFavorite,
          })}
          onClick={this._favoriteCar}
        >
          <i className="fa fa-heart" />
        </div>

        {/* Car make&model */}
        <div className={styles.make}>
          {car.model_year} {car.make}
        </div>
        <div className={styles.model}>
          {car.model} {car.trim}
        </div>

        {/* Mileage */}
        <div className={styles.mileage}>
          <span className="number">{car.mileage.toLocaleString()}</span> Mi.
        </div>

        {/* Payment info */}
        {payment && (
          <div className={styles.payment}>
            <div className={styles.monthyPayment}>
              <label>Monthy Pymt.</label>
              <div className={styles.price}>
                ${payment.monthly_payment_cents / 100}
              </div>
            </div>
            <div className={styles.startPayment}>
              <label>Start Pymt.</label>
              <div className={styles.price}>
                ${payment.start_fee_cents / 100}
              </div>
            </div>
          </div>
        )}

        {/* Color info */}
        <div className={styles.colorInfo}>
          <div className={styles.exterior}>
            <label>Exterior Color</label>
            <div className={styles.color}>Metallic</div>
          </div>
          <div className={styles.interior}>
            <label>Interior Color</label>
            <div className={styles.color}>Black</div>
          </div>
        </div>
      </div>
    )
  }
}

CarDetail.propTypes = {
  car: PropTypes.object.isRequired,
  favoriteCarAction: PropTypes.func.isRequired,
}

export default CarDetail
