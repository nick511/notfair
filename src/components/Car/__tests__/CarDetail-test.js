import React from 'react'
import { shallow } from 'enzyme'

import CarDetail from '../CarDetail'

describe('<CarDetail />', () => {
  let wrapper = null
  const car = {
    id: '19XFC2F59GE2276732016',
    make: 'Honda',
    mileage: 35292,
    model: 'Civic Sedan',
    model_year: '2016',
    product_financials: [
      {
        id: 19882216,
        monthly_payment_cents: 29000,
        start_fee_cents: 90000,
      },
    ],
    trim: 'LX',
    isFavorite: false,
  }
  const favoriteCarAction = jest.fn()

  beforeEach(() => {
    wrapper = shallow(
      <CarDetail car={car} favoriteCarAction={favoriteCarAction} />,
    )
  })

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders car make, mileage and prices', () => {
    // make&model
    expect(wrapper.find('.make').text()).toEqual(
      `${car.model_year} ${car.make}`,
    )
    expect(wrapper.find('.model').text()).toEqual(`${car.model} ${car.trim}`)

    // mileage
    expect(wrapper.find('.mileage .number').text()).toEqual(
      car.mileage.toLocaleString(),
    )

    // prices
    expect(wrapper.find('.monthyPayment .price').text()).toEqual(
      '$' + car.product_financials[0].monthly_payment_cents / 100,
    )
    expect(wrapper.find('.startPayment .price').text()).toEqual(
      '$' + car.product_financials[0].start_fee_cents / 100,
    )
  })

  it('chages favorite icon style after click the icon', () => {
    const action = () => {
      wrapper.setProps({
        car: { ...car, isFavorite: true },
      })
    }
    wrapper = shallow(<CarDetail car={car} favoriteCarAction={action} />)

    // Should be gray
    expect(wrapper.find('.favorite').hasClass('favorited')).toEqual(false)

    // Should has color
    wrapper.find('.favorite').simulate('click')
    expect(wrapper.find('.favorite').hasClass('favorited')).toEqual(true)
  })
})
