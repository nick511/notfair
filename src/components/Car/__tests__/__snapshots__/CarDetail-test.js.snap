// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarDetail /> renders without crashing 1`] = `
<div
  className="component"
>
  <div
    className="favorite favorite"
    onClick={[Function]}
  >
    <i
      className="fa fa-heart"
    />
  </div>
  <div
    className="make"
  >
    2016
     
    Honda
  </div>
  <div
    className="model"
  >
    Civic Sedan
     
    LX
  </div>
  <div
    className="mileage"
  >
    <span
      className="number"
    >
      35,292
    </span>
     Mi.
  </div>
  <div
    className="payment"
  >
    <div
      className="monthyPayment"
    >
      <label>
        Monthy Pymt.
      </label>
      <div
        className="price"
      >
        $
        290
      </div>
    </div>
    <div
      className="startPayment"
    >
      <label>
        Start Pymt.
      </label>
      <div
        className="price"
      >
        $
        900
      </div>
    </div>
  </div>
  <div
    className="colorInfo"
  >
    <div
      className="exterior"
    >
      <label>
        Exterior Color
      </label>
      <div
        className="color"
      >
        Metallic
      </div>
    </div>
    <div
      className="interior"
    >
      <label>
        Interior Color
      </label>
      <div
        className="color"
      >
        Black
      </div>
    </div>
  </div>
</div>
`;
